runtime/compiler/trj9/optimizer/ValueJProfiling.cpp

```diff
@@ -0,0 +1,883 @@
+/*******************************************************************************
+ * Copyright (c) 2000, 2017 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *******************************************************************************/
+#include "ValueJProfiling.hpp"
+
+#include "il/Block.hpp"
+#include "infra/Cfg.hpp"
+#include "infra/TRCfgEdge.hpp"
+#include "infra/Checklist.hpp"
+#include "infra/ILWalk.hpp"
+#include "infra/List.hpp"
+#include "optimizer/Optimization_inlines.hpp"
+#include "il/Node_inlines.hpp"
+#include "infra/Checklist.hpp"             // for TR::NodeChecklist
+#include "ras/DebugCounter.hpp"
+#include "runtime/J9Profiler.hpp"
+#include "control/Recompilation.hpp"              // for TR_Recompilation, etc
+#include "control/RecompilationInfo.hpp"              // for TR_Recompilation, etc
+#include "codegen/CodeGenerator.hpp"
+#include "optimizer/TransformUtil.hpp"            // for TransformUtil
+
+#define COLD_BLOCK_FREQ 6
+
+// There are existing versions of these mappings, however, they will round up
+// types smaller than Int32
+TR::ILOpCodes
+directStore(TR::DataType dt)
```

Comment:
The key here is the note about the round up - in Java sub-int sized values will be widened to int - this needs the specific width. This could potentially go somewhere else, but I am fine with it being here since it is an anomaly. I don't think we should be reaching up to OMR directly rather than going via the TR mapping - it is a dangerous precedent and leaves this code open to being broken by changes in OMR that could be completely unreleated. Java chose to replace the OMR definitions so this code needs to handle things its own way IMO. When we go to contribute this up to OMR then we can revisit the design.

Reply:
Doxygen on these methods and the reason for their split from OMR is necessary.

Reply:
Doxygen comments have been added to these methods.